src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,487 @@
+                MutableCharSpan label(buffer);
+                err1 = d->GetModeLabelByIndex(i, label);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+                mode.label = label;
+
+                // Get the mode value
+                err1 = d->GetModeValueByIndex(i, mode.mode);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+
+                // Get the mode tags
+                ModeTagStructType tagsBuffer[8];
+                DataModel::List<ModeTagStructType> tags(tagsBuffer);
+                err1 = d->GetModeTagsByIndex(i, tags);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+                mode.modeTags = tags;
+
+                ReturnErrorOnFailure(encoder.Encode(mode));
+            }
+            return CHIP_NO_ERROR;
+        });
+        ReturnErrorOnFailure(err);
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+// Implements checking before attribute writes.
+CHIP_ERROR Instance::Write(const ConcreteDataAttributePath & attributePath, AttributeValueDecoder & aDecoder)
+{
+    DataModel::Nullable<uint8_t> newMode;
+    ReturnErrorOnFailure(aDecoder.Decode(newMode));
+    Status status;
+
+    if (!newMode.IsNull()) // This indicates that the new mode is null.
+    {
+        if (!IsSupportedMode(newMode.Value()))
+        {
+            return StatusIB(Protocols::InteractionModel::Status::InvalidCommand).ToChipError();
+        }
+    }
+
+    switch (attributePath.mAttributeId)
+    {
+    case ModeBase::Attributes::StartUpMode::Id:
+        status = UpdateStartUpMode(newMode);
+        return StatusIB(status).ToChipError();
+    case ModeBase::Attributes::OnMode::Id:
+        status = UpdateOnMode(newMode);
+        return StatusIB(status).ToChipError();
+    }
+
+    return StatusIB(Protocols::InteractionModel::Status::InvalidCommand).ToChipError();
+}
+
+Status Instance::UpdateCurrentMode(uint8_t aNewMode)
+{
+    if (!IsSupportedMode(aNewMode))
+    {
+        return Protocols::InteractionModel::Status::InvalidValue;
+    }
+    uint8_t oldMode = mCurrentMode;
+    mCurrentMode    = aNewMode;
+    if (mCurrentMode != oldMode)
+    {
+        // Write new value to persistent storage.
+        GetAttributePersistenceProvider()->WriteScalarValue(
+            ConcreteAttributePath(mEndpointId, mClusterId, Attributes::CurrentMode::Id), mCurrentMode);
+        // The Administrator Commissioning cluster is always on the root endpoint.
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::CurrentMode::Id);
+    }
+    return Protocols::InteractionModel::Status::Success;
+}
+
+Status Instance::UpdateStartUpMode(DataModel::Nullable<uint8_t> aNewStartUpMode)
+{
+    if (!aNewStartUpMode.IsNull())
+    {
+        if (!IsSupportedMode(aNewStartUpMode.Value()))
```

Comment:
Pretty much all the UpdateCurrentMode comments apply here.

Reply:
Done.